### DISCUSSION TITLE
[개선안] 공동 모델 구조체 저장소 운영

### DISCUSSION THREAD

**Discussion by niconicodex** [cloud-barista/cm-honeybee#3] [Ideas]
## 공통 모델 구조체 저장소 운영 방안

마이그레이션을 위한 정보 수집을 가장 초반에 이루어 지며, 이후 수집된 데이터들을 기반으로 프레임워크들 간에 상호작용이 있을 것 같습니다.
데이터를 처리하는 과정에서 필요한 각 모델의 구조체를 공통 관리하면, 재 사용성/일관성/유지보수 용이성 등에 대한 보증이 될 것 같습니다.

위 같은 사유로 공통 모델 구조체를 관리하는 저장소를 운영하는 것도 좋은 방법인 것  같습니다.

현재는 저장소의 경로 안에 `model/infra` 디렉토리에 구현 되어 있습니다.
[[REFACTOR] splite out the model structure-f7a9c19](https://github.com/cloud-barista/cm-honeybee/commit/f7a9c1931eef76b78fc208ba6bb15612322e2e44) 참고 부탁드립니다.

### 사례 참고
이전 수행했던 과제에서는 gRPC Proto 파일 관리와 공통 모델 구조체 관리를 위한 저장소를 두 개 운영 하였습니다.

* gRPC repo : https://github.com/hcloud-classic/hcc-melody
* Common Models repo : https://github.com/hcloud-classic/hcc-model

@powerkimhub  @yunkon-kim 위 내용 검토 부탁드립니다.

**Comment by powerkimhub** [comment]
- 좋은 제안이십니다.
- 다만, 초기에는 검토나 수정이 빈번할 수 있는 데 
  - 이 시기에는 제 3자가 모델에 대한 이해와 편집이 쉬운 
  - 엑셀이나 ppt 등의 형태로도 공유가 되면 좋을 것 같습니다.

**Reply by niconicodex** [reply]
넵! [(HoneyBee) 소스 컴퓨팅 환경별 레퍼런스 아키텍처 및 분석 메트릭 정의 공유](https://docs.google.com/spreadsheets/d/1TCw8tbJdDPfBaYB6JPoXh5YUjTbKgpIr) 를 최대한 활용해서 작성해 보겠습니다.

**Comment by yunkon-kim** [comment]
좋은 방안을 말씀해주셔서 감사합니다 ^^

(Model 관련)
말씀해주신 방안은 지난번 컨소시엄 회의에서 다루어진 안건이기도하며,
현재 [CM-Damselfly](https://github.com/cloud-barista/cm-damselfly)와 직접적인 관련성이 있습니다(CC. @innodreamer).

따라서, 빠른 시일 내에 상호 논의를 진행하는 것이 좋을 것 같습니다.

(gRPC 관련)
gRPC proto를 통한 model 및 function/lib/API (즉, RPC 관련)를 제공하는 것에는 분명한 장점이 있음을 충분히 공감하고 있습니다. 하지만, 기존 gRPC와 RESTful API 모두 제공하던 것에서 RESTful API만 제공하는 것으로 커뮤니티 차원의 결정이 있었습니다.

따라서, gRPC proto 및 ProtoBuf 관련 사항은 별개로 논의하는 것이 좋을 것 같습니다. 
(예를 들어, Server streaming, Client Streaming, BI Directional Streaming 가 필요한 케이스 등에 대한 논의)
